### PR TITLE
Improve distribution detection

### PR DESCRIPTION
**Pull request by suraia**
This PR improves distribution detection in two ways:
1. Lowercase the distribution name for consistency. Currently, there is "fedora" and "Ubuntu".
2. Support Ubuntu's YY.MM version scheme. Version 14.04 is not equivalent to version 14.10.

Before:

```
$ spack arch
linux-Ubuntu14-x86_64
```

After:

```
$ spack arch
linux-ubuntu14.04-x86_64
```

### CODE REVIEW

**Comment by eschnett** [comment]
Fundamental changes like this that force people to rebuild all packages should be queued up, and should be applied together, on a flag day once a month.

They should also be thoroughly tested, e.g. by building about one hundred packages on ten different systems, including OS X and Cray.

Otherwise I'll be forced to rebuild everything, but I can't since there might be a problem (reindex? lmod?), and I'm left with an unfixable mess, except if I go back to a known-to-be working version of Spack.

Alternatively, there could be a backward compatibility mechanism that e.g. automatically installs the respective symlinks to make existing packages continue to work.

**Comment by tgamblin** [comment]
I created a label for issues like this (spec-change).  I like the idea of lowercasing the distro name -- I think it's a good idea.  I also think we could hold off as @eschnett suggests and only merge things like this on certain days.  I'll work on a schedule for that.

**Comment by suraia** [comment]
Sorry, I accidentally deleted the wrong branch...

**Comment by suraia** [comment]
@eschnett @tgamblin Any updates regarding this? Our Ubuntu LTS installation is currently misidentified (14 instead of 14.04) because of this.

**Comment by davydden** [comment]
my understanding is that https://github.com/LLNL/spack/pull/1629 will most likely be adopted.

**Comment by suraia** [comment]
I just tried the changes from #1629 and while they do fix the case inconsistency ("Ubuntu" becomes "ubuntu"), they do not fix the version problem (Ubuntu 14.04 is still identified as "ubuntu14").

**Comment by davydden** [comment]
>  they do not fix the version problem (Ubuntu 14.04 is still identified as "ubuntu14").

This may not be perceived as a problem by everyone. Let's wait fora  feedback from maintainers.

**Comment by tgamblin** [comment]
@suraia: The idea with the OS identifiers in Spack is that they're sufficient for binary compatibility between OS versions, but not so specific as to require a bunch of rebuilds when it's not necessary.  I don't think going down to a patch version on Ubuntu is actually the greatest idea, because it'll mean you have to rebuild everything even on minor OS upgrades.  I think the full provenance would probably be good to keep around, but as this is it affects the Spec.

So, basically, I want better distro detection but it probably needs to be coarser than this.

Thoughts?

**Comment by suraia** [comment]
@tgamblin: Ubuntu releases twice a year and x.04 and x.10 are actually different releases that are not necessarily binary compatible. For example, Ubuntu 16.04 ships with GCC 5.3 while 16.10 will use 6.x. The minor LTS upgrades are actually called 16.04.x.

I hope that clears up some of the confusion.

**Comment by tgamblin** [comment]
Closed in favor of #1629.
